src/oscore/oscore_context.c

```diff
@@ -0,0 +1,787 @@
+    coap_delete_bin_const(association->aad);
+    coap_delete_bin_const(association->nonce);
+    coap_delete_bin_const(association->partial_iv);
+    coap_free(association);
+  }
+}
+
+int
+oscore_new_association(coap_session_t *session,
+                       coap_pdu_t *sent_pdu,
+                       coap_bin_const_t *token,
+                       oscore_recipient_ctx_t *recipient_ctx,
+                       coap_bin_const_t *aad,
+                       coap_bin_const_t *nonce,
+                       coap_bin_const_t *partial_iv,
+                       int is_observe) {
+  oscore_association_t *association;
+
+  association = coap_malloc(sizeof(oscore_association_t));
+  if (association == NULL)
+    return 0;
+
+  memset(association, 0, sizeof(oscore_association_t));
+  association->recipient_ctx = recipient_ctx;
+  association->is_observe = is_observe;
+
+  if (sent_pdu) {
+    size_t size;
+    const uint8_t *data;
+
+    association->sent_pdu = coap_pdu_duplicate(sent_pdu, session,
+                                               token->length, token->s, NULL);
+    if (association->sent_pdu == NULL)
+      goto error;
+    if (coap_get_data(sent_pdu, &size, &data)) {
+      coap_add_data(association->sent_pdu, size, data);
+    }
+  }
+  association->token = coap_new_bin_const(token->s, token->length);
+  if (association->token == NULL)
+    goto error;
+
+  if (aad) {
+    association->aad = coap_new_bin_const(aad->s, aad->length);
+    if (association->aad == NULL)
+      goto error;
+  }
+
+  if (nonce) {
+    association->nonce = coap_new_bin_const(nonce->s, nonce->length);
+    if (association->nonce == NULL)
+      goto error;
+  }
+
+  if (partial_iv) {
+    association->partial_iv =
+        coap_new_bin_const(partial_iv->s, partial_iv->length);
+    if (association->partial_iv == NULL)
+      goto error;
+  }
+
+  OSCORE_ASSOCIATIONS_ADD(session->associations, association);
+  return 1;
+
+error:
+  oscore_free_association(association);
+  return 0;
+}
+
+oscore_association_t *
+oscore_find_association(coap_session_t *session, coap_bin_const_t *token) {
+  oscore_association_t *association;
+
+  OSCORE_ASSOCIATIONS_FIND(session->associations, token, association);
+  return association;
+}
+
+int
+oscore_delete_association(coap_session_t *session,
+                          oscore_association_t *association) {
+  if (association) {
+    OSCORE_ASSOCIATIONS_DELETE(session->associations, association);
+    oscore_free_association(association);
+    return 1;
+  }
+  return 0;
+}
+
+void
+oscore_delete_server_associations(coap_session_t *session) {
+  if (session) {
+    oscore_association_t *association;
+    oscore_association_t *tmp;
+
+    OSCORE_ASSOCIATIONS_ITER_SAFE(session->associations, association, tmp) {
+      OSCORE_ASSOCIATIONS_DELETE(session->associations, association);
+      oscore_free_association(association);
+    }
+    session->associations = NULL;
+  }
+}
+
+int
```

Comment:
This function is never called, so I suggest removing it.

Reply:
Removed